[feat] - Refactor AzureSD

heliapb opened the pull request:
## Description

From issue https://github.com/prometheus-operator/prometheus-operator/issues/7202

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [x] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
<!-- How you tested it? How do you know it works? -->
Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Add metadataConfig field to the Prometheus CRD for configuring how remote-write sends metadata information.
- Generate correct scraping configuration for Probes with empty or unset module parameter.
-->

```release-note
ENHANCEMENT: Refactor AzureSD
```

simonpasquier reviewed (dismissed):
thanks!

slashpai commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -692,7 +692,7 @@ type AzureSDConfig struct {
 	// instead be specified in the relabeling rule.
 	// +optional
-	Port *int `json:"port"`
+	Port *int `json:"port,omitempty"`
```
I think we can include minimum and maximum value for port number like did in EC2SDConfig
cc: @mviswanathsai

heliapb replied:
thanks, fixed

slashpai commented:
A few more fields still need minlength validation, `// +kubebuilder:validation:MinLength=1`

```go
        // Optional tenant ID. Only required with the OAuth authentication method.
	// +optional
	TenantID *string `json:"tenantID,omitempty"`
	// Optional client ID. Only required with the OAuth authentication method.
	// +optional
	ClientID *string `json:"clientID,omitempty"`
	// Optional resource group name. Limits discovery to this resource group.
	// +optional
	ResourceGroup *string `json:"resourceGroup,omitempty"`
```

mviswanathsai commented:
Adding to what Jayapriya said,
```
	// The Azure environment.
        // +kubebuilder:validation:MinLength=1
	// +optional
	Environment *string `json:"environment,omitempty"`
```

mviswanathsai commented:
```
AuthenticationMethod *AuthenticationMethodType `json...`
```
And 
```
// +kubebuilder:validation:Enum=OAuth;ManagedIdentity;SDK
type AuthenticationMethodType string

const (
AuthMethodTypeOAuth AuthenticationMethodType = "OAuth"
// and so on, for the rest of the types
)
```
This makes the code more reusable, we can use these constants in our tests instead of using strings and risking typos. Similar to what we have with `DNSRecordTypes`.

mviswanathsai commented:
Just as a heads up, looking at the `UnmarshalYAML()` function of each of the service discoveries gives a good idea of what is expected from each field. In this case, https://github.com/prometheus/prometheus/blob/main/discovery/azure/azure.go#L146

mviswanathsai commented on test/e2e/scrapeconfig_test.go:
```diff
@@ -2035,3 +2038,107 @@ var IonosSDTestCases = []scrapeCRDTestCase{
+			},
+		},
+		expectedError: true,
+	},
```
Could you also add tests for the other optional fields, checking for the correct working of the min length length validation?

mviswanathsai replied:
Also, once the version checks are added I would also like to see some versioned testcases to see that the object is being rejected, if it defines some field that is not yet supported by the cluster's Prometheus version.

mviswanathsai commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -675,24 +675,29 @@ type AzureSDConfig struct {
 	// Optional client secret. Only required with the OAuth authentication method.
 	// +optional
 	ClientSecret *corev1.SecretKeySelector `json:"clientSecret,omitempty"`
 	// Optional resource group name. Limits discovery to this resource group.
```
```suggestion
	// Optional resource group name. Limits discovery to this resource group.
	// Requires  Prometheus v2.35.0 and above
```
Also, we need to check for the version and reject any object that fails this check at `resource_selector.go`

mviswanathsai commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -675,24 +675,29 @@ type AzureSDConfig struct {
+	// +kubebuilder:validation:Maximum=65535
 	// +optional
-	Port *int `json:"port"`
+	Port *int32 `json:"port,omitempty"`
```
Azure also supports HTTP client configuration. See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#azure_sd_config

You would need to add something similar to:
```Go
	// BasicAuth information to authenticate against the target HTTP endpoint.
	// More info: https://prometheus.io/docs/operating/configuration/#endpoints
	// Cannot be set at the same time as `authorization`, or `oAuth2`.
	// +optional
	BasicAuth *v1.BasicAuth `json:"basicAuth,omitempty"`
	// Authorization header configuration to authenticate against the target HTTP endpoint.
	// Cannot be set at the same time as `oAuth2`, or `basicAuth`.
	// +optional
	Authorization *v1.SafeAuthorization `json:"authorization,omitempty"`
	// Optional OAuth 2.0 configuration to authenticate against the target HTTP endpoint.
	// Cannot be set at the same time as `authorization`, or `basicAuth`.
	// +optional
	OAuth2         *v1.OAuth2 `json:"oauth2,omitempty"`
	v1.ProxyConfig `json:",inline"`
	// TLS configuration applying to the target HTTP endpoint.
	// +optional
	TLSConfig *v1.SafeTLSConfig `json:"tlsConfig,omitempty"`
	// Configure whether HTTP requests follow HTTP 3xx redirects.
	// +optional
	FollowRedirects *bool `json:"followRedirects,omitempty"`
	// Whether to enable HTTP2.
	// +optional
	EnableHTTP2 *bool `json:"enableHTTP2,omitempty"`
```

mviswanathsai replied:
HTTP client config for AzureSD is supported only for Prometheus >= v2.31.0. So I would suggest we also add the rejection logic at `resource_selector.go` for this.

heliapb replied:
Added thanks

mviswanathsai commented:
I think you would also need to do something similar to the change in #7239

mviswanathsai reviewed (changes requested):
Hey @heliapb !
Thank you for addressing my previous feedback in your PR. However, I would like to reiterate my main concern regarding the structure of the e2e test cases. Specifically, I believe we can make the test cases more granular and verbose to improve clarity and maintainability. 

Instead of grouping all the empty or invalid fields into a single test case, I recommend breaking them down into individual test cases for each field. This ensures each test case is specific and focuses on validating one scenario at a time. This approach makes debugging easier and ensures edge cases are covered comprehensively.

For instance, for a field like `SubscriptionID`, we should have:
1. A test where it is correctly defined.
2. A test where it is missing.
3. A test where it is invalid (e.g., an empty string, invalid UUID).

Below is an example illustrating the difference:

### Current Test Case
```go
{
	name: "Invalid AzureSD Missing SubscriptionID",
	scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
		AzureSDConfigs: []monitoringv1alpha1.AzureSDConfig{
			{
				Environment:          ptr.To("AzurePublicCloud"),
				AuthenticationMethod: ptr.To(monitoringv1alpha1.AuthMethodTypeSDK),
				TenantID:             ptr.To("22222222-2222-2222-2222-222222222222"),
			},
		},
	},
	expectedError: true,
},
```

### Suggested Granular Tests
```go
// Test where SubscriptionID is correctly defined
{
	name: "Valid AzureSD With SubscriptionID",
	scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
		AzureSDConfigs: []monitoringv1alpha1.AzureSDConfig{
			{
				SubscriptionID: "11111111-1111-1111-1111-111111111111",
			},
		},
	},
	expectedError: false,
},

// Test where SubscriptionID is missing
{
	name: "Invalid AzureSD Missing SubscriptionID",
	scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
		AzureSDConfigs: []monitoringv1alpha1.AzureSDConfig{
			{
				Environment:          ptr.To("AzurePublicCloud"),
				AuthenticationMethod: ptr.To(monitoringv1alpha1.AuthMethodTypeSDK),
			},
		},
	},
	expectedError: true,
},

// Test where SubscriptionID is invalid. In this case, empty.
{
	name: "Invalid AzureSD With Invalid SubscriptionID",
	scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
		AzureSDConfigs: []monitoringv1alpha1.AzureSDConfig{
			{
				SubscriptionID: "",
			},
		},
	},
	expectedError: true,
},
```

Its a bit more work, but as I said, it stays in line with how we have defined our other testcases. There are a few unresolved comments that seem to have missed your attention, please take a look at them as well.

heliapb commented:
Hi @mviswanathsai sorry I've had missed your last comments, fixed those. Also added the more granular tests, let me know if anything else needs fix, and thanks for the feedback, much appreciated.

heliapb commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -656,43 +656,86 @@ type EC2SDConfig struct {
 	// Optional resource group name. Limits discovery to this resource group.
+	// Requires  Prometheus v2.35.0 and above
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MinLength=1
```
Sorry I had missed those reviews, fixed

heliapb commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -656,43 +656,86 @@ type EC2SDConfig struct {
 	TenantID *string `json:"tenantID,omitempty"`
 	// Optional client ID. Only required with the OAuth authentication method.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MinLength=1
```
Sorry I had missed those reviews, fixed

heliapb commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -656,43 +656,86 @@ type EC2SDConfig struct {
 	// +required
 	SubscriptionID string `json:"subscriptionID"`
 	// Optional tenant ID. Only required with the OAuth authentication method.
+	// +kubebuilder:validation:MinLength=1
```
Sorry it was accidental, fixed thanks

mviswanathsai reviewed (approved):
LGTM.
CC: @slashpai

slashpai commented:
@heliapb could you please rebase your PR, after GCE PR merge there is conflict

heliapb commented:
Hi @slashpai rebase done, thanks

slashpai reviewed (approved):
Great work, Thank you @heliapb and @mviswanathsai  :)
